<a href="https://colab.research.google.com/github/nidhicodes4045/datascience442/blob/main/Malik's_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import time
t1 = time.perf_counter()
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import itertools
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

df = pd.read_csv("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/Malik's_EDA_Datasets/eda_df.csv")

In [ ]:
df.head()

County  Population State     mergedkeycol  \
0  autaugacounty       54571    al  autaugacountyal   
1  baldwincounty      182265    al  baldwincountyal   
2  barbourcounty       27457    al  barbourcountyal   
3     bibbcounty       22915    al     bibbcountyal   
4   blountcounty       57322    al   blountcountyal   

   Low Access Numbers.Low Income People.1/2 Mile  \
0                                          15518   
1                                          48117   
2                                          11901   
3                                           8349   
4                                          19806   

   Low Access Numbers.People.1/2 Mile  Percentage of Low Income People  \
0                               49497                         0.313514   
1                              165616                         0.290534   
2                               23762                         0.500842   
3                               20989                         0.397780   
4                               54933                         0.360548   

   Education.Bachelor's Degree or Higher  Education.High School or Higher  \
0                                   26.6                             88.5   
1                                   31.9                             90.8   
2                                   11.6                             73.2   
3                                   10.4                             79.1   
4                                   13.1                             80.5   

   Ethnicities.Black Alone  ...  Housing.Homeownership Rate  \
0                     20.1  ...                        73.3   
1                      8.8  ...                        75.2   
2                     48.2  ...                        60.9   
3                     21.3  ...                        74.4   
4                      1.7  ...                        78.8   

   Housing.Median Value of Owner-Occupied Units  \
0                                        154500   
1                                        197900   
2                                         90700   
3                                         92800   
4                                        127800   

   Income.Median Houseold Income  Miscellaneous.Foreign Born  \
0                          58731                         2.3   
1                          58320                         3.7   
2                          32525                         2.7   
3                          47542                         1.5   
4                          49358                         4.5   

   Miscellaneous.Language Other than English at Home  \
0                                                4.3   
1                                                5.4   
2                                                7.1   
3                                                2.3   
4                                                8.2   

   Population.Population per Square Mile  Employment.Firms.Minority-Owned  \
0                                   91.8                              616   
1                                  114.6                             2130   
2                                   31.0                              657   
3                                   36.8                              205   
4                                   88.9                              368   

   Employment.Firms.Nonminority-Owned  Employment.Firms.Total  \
0                                2160                    2949   
1                               16893                   19732   
2                                 941                    1687   
3                                1050                    1310   
4                                4011                    4550   

   Percentage of Minority Businesses  
0                           0.208884  
1                           0.107946  
2                           0.389449  
3                           0.156489  
4

# Dropping columns

In [ ]:
df["Percentage of Nonminority Businesses"] = df["Employment.Firms.Nonminority-Owned"] / df["Employment.Firms.Total"]
df.fillna(0, inplace=True)
df.drop(['County', 'Population', 'State', 'mergedkeycol'], axis = 1, inplace = True)
df.drop(['Low Access Numbers.Low Income People.1/2 Mile', 'Low Access Numbers.People.1/2 Mile', 'Employment.Firms.Minority-Owned', 'Population.Population per Square Mile', 'Employment.Firms.Nonminority-Owned', 'Employment.Firms.Total'], axis = 1, inplace = True)
df.head()

Percentage of Low Income People  Education.Bachelor's Degree or Higher  \
0                         0.313514                                   26.6   
1                         0.290534                                   31.9   
2                         0.500842                                   11.6   
3                         0.397780                                   10.4   
4                         0.360548                                   13.1   

   Education.High School or Higher  Ethnicities.Black Alone  \
0                             88.5                     20.1   
1                             90.8                      8.8   
2                             73.2                     48.2   
3                             79.1                     21.3   
4                             80.5                      1.7   

   Ethnicities.Hispanic or Latino  \
0                             3.0   
1                             4.7   
2                             4.5   
3                             2.8   
4                             9.7   

   Ethnicities.White Alone\t not Hispanic or Latino  Ethnicities.Asian Alone  \
0                                              73.8                      1.2   
1                                              83.2                      1.1   
2                                              45.5                      0.5   
3                                              74.4                      0.2   
4                                              86.8                      0.3   

   Housing.Homeownership Rate  Housing.Median Value of Owner-Occupied Units  \
0                        73.3                                        154500   
1                        75.2                                        197900   
2                        60.9                                         90700   
3                        74.4                                         92800   
4                        78.8                                        127800   

   Income.Median Houseold Income  Miscellaneous.Foreign Born  \
0                          58731                         2.3   
1                          58320                         3.7   
2                          32525                         2.7   
3                          47542                         1.5   
4                          49358                         4.5   

   Miscellaneous.Language Other than English at Home  \
0                                                4.3   
1                                                5.4   
2                                                7.1   
3                                                2.3   
4                                                8.2   

   Percentage of Minority Businesses  Percentage of Nonminority Businesses  
0                           0.208884                              0.732452  
1                           0.107946                              0.856122  
2                           0.389449                              0.557795  
3                           0.156489                              0.801527  
4                           0.080879                              0.881538

# Renaming Columns


In [ ]:
df.rename(columns={
    'Percentage of Low Income People': 'Low_Income_Pct',
    'Education.Bachelor\'s Degree or Higher': 'Bachelors_Pct',
    'Education.High School or Higher': 'High_School_Pct',
    'Ethnicities.Black Alone': 'Black_Pct',
    'Ethnicities.Hispanic or Latino': 'Hispanic_Pct',
    'Ethnicities.White Alone\t not Hispanic or Latino': 'White_Non_Hispanic_Pct',
    'Ethnicities.Asian Alone': 'Asian_Pct',
    'Housing.Homeownership Rate': 'Homeownership_Rate',
    'Housing.Median Value of Owner-Occupied Units': 'Median_Home_Value',
    'Income.Median Houseold Income': 'Median_Household_Income',
    'Miscellaneous.Foreign Born': 'Foreign_Born_Pct',
    'Miscellaneous.Language Other than English at Home': 'Lang_Other_Than_English_Pct',
    'Percentage of Minority Businesses': 'Minority_Business_Pct',
    'Percentage of Nonminority Businesses': 'Nonminority_Business_Pct'
}, inplace=True)
df.head()

Low_Income_Pct  Bachelors_Pct  High_School_Pct  Black_Pct  Hispanic_Pct  \
0        0.313514           26.6             88.5       20.1           3.0   
1        0.290534           31.9             90.8        8.8           4.7   
2        0.500842           11.6             73.2       48.2           4.5   
3        0.397780           10.4             79.1       21.3           2.8   
4        0.360548           13.1             80.5        1.7           9.7   

   White_Non_Hispanic_Pct  Asian_Pct  Homeownership_Rate  Median_Home_Value  \
0                    73.8        1.2                73.3             154500   
1                    83.2        1.1                75.2             197900   
2                    45.5        0.5                60.9              90700   
3                    74.4        0.2                74.4              92800   
4                    86.8        0.3                78.8             127800   

   Median_Household_Income  Foreign_Born_Pct  Lang_Other_Than_English_Pct  \
0                    58731               2.3                          4.3   
1                    58320               3.7                          5.4   
2                    32525               2.7                          7.1   
3                    47542               1.5                          2.3   
4                    49358               4.5                          8.2   

   Minority_Business_Pct  Nonminority_Business_Pct  
0               0.208884                  0.732452  
1               0.107946                  0.856122  
2               0.389449                  0.557795  
3               0.156489                  0.801527  
4               0.080879                  0.881538

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3138 entries, 0 to 3137
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Low_Income_Pct               3138 non-null   float64
 1   Bachelors_Pct                3138 non-null   float64
 2   High_School_Pct              3138 non-null   float64
 3   Black_Pct                    3138 non-null   float64
 4   Hispanic_Pct                 3138 non-null   float64
 5   White_Non_Hispanic_Pct       3138 non-null   float64
 6   Asian_Pct                    3138 non-null   float64
 7   Homeownership_Rate           3138 non-null   float64
 8   Median_Home_Value            3138 non-null   int64  
 9   Median_Household_Income      3138 non-null   int64  
 10  Foreign_Born_Pct             3138 non-null   float64
 11  Lang_Other_Than_English_Pct  3138 non-null   float64
 12  Minority_Business_Pct        3138 non-null   float64
 13  Nonminority_Busine

# No specific feature selection linear regression

In [ ]:
# separate the response variable and rest of variables as features
features = [
    "Bachelors_Pct",
    "High_School_Pct",
    "Black_Pct",
    "Hispanic_Pct",
    "White_Non_Hispanic_Pct",
    "Asian_Pct",
    "Homeownership_Rate",
    "Median_Home_Value",
    "Median_Household_Income",
    "Foreign_Born_Pct",
    "Lang_Other_Than_English_Pct",
    "Minority_Business_Pct",
    "Nonminority_Business_Pct"
]

X = df[features].values
y = df['Low_Income_Pct'].values
# Define the independent variables (features, x) and the dependent variable (target, y)
# each row in df is an observation
# predictors are the input variables that may influence the target variable

# 20% Test Size (600 rows): gives about 2400 training rows and 600 testing rows
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and fit the model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
# MSE calculates the avg sqr diff between the actual and predicted values
# lower MSE means better fit

r2 = r2_score(y_test, y_pred)
print(f"R-squared: {r2}")
# r-sqr value represents the proportion of the variance in the dependent variable
# that is predictable from the independent variables, the higher the better fit

# Print the coefficients
print(f"Intercept: {model.intercept_}")
print(f"Slope: {model.coef_}")


Mean Squared Error: 0.0014638361143351078
R-squared: 0.86030165370141
Intercept: 1.234628231418692
Slope: [ 1.07094190e-03 -5.42376652e-03 -1.66106844e-03 -1.61253721e-03
 -2.09367407e-03 -3.30034835e-04 -2.73108262e-04  1.18909652e-07
 -5.65644648e-06 -6.53263362e-04 -8.20271268e-04  9.88088266e-02
  7.14141956e-02]


## Interpreting the output

Mean Squared Error (MSE): 0.0014638361143351078
*   This value represents the average of the squared differences between the actual "Low_Income_Pct" values in your test set (y_test) and the values predicted by your linear regression model (y_pred).
*   A lower MSE indicates a better fit of the model to the data, as the prediction errors are smaller on average.
*   A value of 0.00146 suggests that, on average, the squared difference between your model's predictions and the actual low-income percentages in the test set is quite small.

R-squared: 0.86030165370141
*   An R-squared value closer to 1 indicates a better fit, suggesting that your model explains a large portion of the variance in the dependent variable. 0.86 is generally considered a good R-squared value, indicating a strong relationship between the features and the target variable.

Coefficients (Slope)

Bachelors_Pct                | 1.07094190e-03 (0.0011) | For every 1 percentage point increase in the percentage of the population with a Bachelor's degree, the predicted "Low_Income_Pct" increases by approximately 0.0011 units (holding other features constant). This suggests a very small positive relationship.                                                                                                                                      
High_School_Pct             | -5.42376652e-03 (-0.0054) | For every 1 percentage point increase in the percentage of the population with a high school degree, the predicted "Low_Income_Pct" decreases by approximately 0.0054 units (holding other features constant). This suggests a small negative relationship.                                                                                                                                       
Black_Pct                    | -1.66106844e-03 (-0.0017) | For every 1 percentage point increase in the percentage of the population that is Black, the predicted "Low_Income_Pct" decreases by approximately 0.0017 units (holding other features constant). This suggests a small negative relationship.                                                                                                                                        
Hispanic_Pct                 | -1.61253721e-03 (-0.0016) | For every 1 percentage point increase in the percentage of the population that is Hispanic, the predicted "Low_Income_Pct" decreases by approximately 0.0016 units (holding other features constant). This suggests a small negative relationship.                                                                                                                                    
White_Non_Hispanic_Pct    | -2.09367407e-03 (-0.0021) | For every 1 percentage point increase in the percentage of the population that is White (non-Hispanic), the predicted "Low_Income_Pct" decreases by approximately 0.0021 units (holding other features constant). This suggests a small negative relationship.                                                                                                                            
Asian_Pct                    | -3.30034835e-04 (-0.0003) | For every 1 percentage point increase in the percentage of the population that is Asian, the predicted "Low_Income_Pct" decreases by approximately 0.0003 units (holding other features constant). This suggests a very small negative relationship.                                                                                                                                        
Homeownership_Rate           | -2.73108262e-04 (-0.0003) | For every 1 percentage point increase in the homeownership rate, the predicted "Low_Income_Pct" decreases by approximately 0.0003 units (holding other features constant). This suggests a very small negative relationship.                                                                                                                                                              
Median_Home_Value           | 1.18909652e-07 (0.00000012) | For every one-unit increase in the median home value, the predicted "Low_Income_Pct" increases by approximately 0.00000012 units (holding other features constant). This suggests a very, very small positive relationship. You might need to consider the scale of "Median_Home_Value" to understand the practical significance.                                                              

Median_Household_Income                | -5.65644648e-06 (-0.0000057) | For every one-unit increase in the median income, the predicted "Low_Income_Pct" decreases by approximately 0.0000057 units (holding other features constant). This suggests a very small negative relationship. Again, consider the scale of "Median_Household_Income".                                                                                                                                     
Foreign_Born_Pct            | -6.53263362e-04 (-0.0007) | For every 1 percentage point increase in the percentage of the population that is foreign-born, the predicted "Low_Income_Pct" decreases by approximately 0.0007 units (holding other features constant). This suggests a small negative relationship.                                                                                                                                 
Lang_Other_Than_English_Pct | -8.20271268e-04 (-0.0008) | For every 1 percentage point increase in the percentage of the population that speaks a language other than English at home, the predicted "Low_Income_Pct" decreases by approximately 0.0008 units (holding other features constant). This suggests a small negative relationship.                                                                                                         
Minority_Business_Pct       | 9.88088266e-02 (0.0988)  | For every 1 percentage point increase in the percentage of businesses owned by minorities, the predicted "Low_Income_Pct" increases by approximately 0.0988 units (holding other features constant). This suggests a relatively larger positive relationship compared to most other features.                                                                                                    
Nonminority_Business_Pct    | 7.14141956e-02 (0.0714)  | For every 1 percentage point increase in the percentage of businesses not owned by minorities, the predicted "Low_Income_Pct" increases by approximately 0.0714 units (holding other features constant). This also suggests a relatively larger positive relationship.




## scaling income...

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# Split the data BEFORE scaling to prevent data leakage
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = MinMaxScaler()
# Fit the scaler to the training data and transform it
X_train_scaled = scaler.fit_transform(X_train)
# Transform the test data using the SAME scaler fitted on the training data
X_test_scaled = scaler.transform(X_test)

model_scaled = LinearRegression()
model_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = model_scaled.predict(X_test_scaled)

mse_scaled = mean_squared_error(y_test, y_pred_scaled)
r2_scaled = r2_score(y_test, y_pred_scaled)
print(f"Scaled Mean Squared Error: {mse_scaled}")
print(f"Scaled R-squared: {r2_scaled}")
print(f"Scaled Intercept: {model_scaled.intercept_}")
print(f"Scaled Slope: {model_scaled.coef_}")

Scaled Mean Squared Error: 0.0014638361143345755
Scaled R-squared: 0.8603016537014608
Scaled Intercept: 0.9605061803685891
Scaled Slope: [ 0.08310509 -0.39322307 -0.14368242 -0.15448106 -0.1991084  -0.01412549
 -0.02004615  0.12763762 -0.68327045 -0.03508024 -0.07808982  0.09175374
  0.0714142 ]


### reinterpreting the output

The MSE remains very close to the value before scaling (0.0014638361143351078). This is expected. While the scale of the predictions might change internally during the model fitting process with scaled features, the overall average squared difference between the predicted and actual values of the target variable ("Low_Income_Pct," which is on a 0 to 1 scale) should be consistent. A small MSE still indicates a good fit.

Same with the r2
Model Coefficients (after scaling):

Scaled Slope (Coefficients for each scaled feature):
The coefficients now represent the change in the predicted "Low_Income_Pct" for a one-unit increase in the scaled version of the corresponding feature (which now ranges from 0 to 1).
Bachelors_Pct: An increase of approximately 0.0831 units.

*   Bachelors_Pct: An increase of approximately 0.0831 units.
*   High_School_Pct: A decrease of approximately 0.3932 units, indicating a strong negative relationship across its scaled range.
*   Black_Pct: A decrease of approximately 0.1437 units.
*   Hispanic_Pct: A decrease of approximately 0.1545 units.
*   White_Non_Hispanic_Pct: A decrease of approximately 0.1991 units, suggesting a relatively strong negative relationship across its scaled range.
*   Asian_Pct: A decrease of approximately 0.0141 units, indicating a very small negative relationship relative to its scaled range.
*   Homeownership_Rate: A decrease of approximately 0.0200 units, suggesting a small negative relationship across its scaled range.
*   Median_Home_Value: An increase of approximately 0.1276 units, reflecting a larger impact across the normalized range of home values.
*   Median_Household_Income: A decrease of approximately 0.6833 units, highlighting a much stronger negative relationship across its normalized range.
*   Foreign_Born_Pct: A decrease of approximately 0.0351 units.
*   Lang_Other_Than_English_Pct: A decrease of approximately 0.0781 units.
*   Minority_Business_Pct: An increase of approximately 0.0918 units.
*   Nonminority_Business_Pct: An increase of approximately 0.0714 units.                                                                                                                                                    

# Forward Selection

Continue to scale for consistency

In [ ]:
# forward selection regression

from mlxtend.feature_selection import SequentialFeatureSelector as sfs

lreg = LinearRegression()
sfs_init = sfs(lreg, k_features=5, forward=True, verbose=2, scoring='neg_mean_squared_error')
sfs_init = sfs_init.fit(X_train_scaled, y_train)


[2025-04-11 02:33:41] Features: 1/5 -- score: -0.0031278321815982725
[2025-04-11 02:33:41] Features: 2/5 -- score: -0.002445076602978731
[2025-04-11 02:33:41] Features: 3/5 -- score: -0.0022459271639572584
[2025-04-11 02:33:41] Features: 4/5 -- score: -0.0021461423835735043
[2025-04-11 02:33:42] Features: 5/5 -- score: -0.0020516267322730065

In [ ]:
# print features that were included in the model
feat_names = list(sfs_init.k_feature_names_)
print(feat_names)

['1', '3', '4', '7', '8']


In [ ]:
#map these to feature names
# print features from k_feature_names_
for i in feat_names:
    print(features[int(i)])

High_School_Pct
Hispanic_Pct
White_Non_Hispanic_Pct
Median_Home_Value
Median_Household_Income


In [ ]:
# Create a dataframe with only the selected features
selected_columns = [features[int(i)] for i in feat_names]
df_selected = df[selected_columns]

X_train, X_test, y_train, y_test = train_test_split(df_selected, y,test_size=0.2,random_state=42)

# scale it! but i wont change its name here
# X_train_scaled will refer to the X_train will 13 variables that was scaled initially
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

reg = LinearRegression()
reg.fit(X_train, y_train)

y_pred = reg.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")
print(f"Intercept: {reg.intercept_}")
print(f"Slope: {reg.coef_}")

Mean Squared Error: 0.0015287895405130065
R-squared: 0.8541029500797266
Intercept: 0.841269958804004
Slope: [-0.28399062 -0.09198038 -0.10411682  0.16832357 -0.69023196]


## interpretation of the output

The MSE is slightly higher (0.001528) compared to the MSE before feature selection (0.001463). This indicates that the model with only these 5 selected features has a slightly higher average prediction error on the test data than the model with all 13 original features. However, the difference is quite small.

An R-squared of 0.8541 suggests that approximately 85.41% of the variation in low-income percentage can be explained by the five selected features: "High_School_Pct", "Hispanic_Pct", "White_Non_Hispanic_Pct", "Median_Home_Value", and "Median_Household_Income". This is still a relatively high R-squared value, indicating a good fit despite using fewer features.

High_School_Pct: -0.28399062 - A one-unit increase in the scaled percentage of adults with a high school diploma is associated with a decrease of approximately 0.2840 units in the predicted low-income percentage. This suggests a negative relationship.

Hispanic_Pct: -0.09198038 - A one-unit increase in the scaled percentage of the Hispanic population is associated with a decrease of approximately 0.0920 units in the predicted low-income percentage. This suggests a negative relationship.

White_Non_Hispanic_Pct: -0.10411682 - A one-unit increase in the scaled percentage of the non-Hispanic White population is associated with a decrease of approximately 0.1041 units in the predicted low-income percentage. This suggests a negative relationship.

Median_Home_Value: 0.16832357 - A one-unit increase in the scaled median home value is associated with an increase of approximately 0.1683 units in the predicted low-income percentage. This suggests a positive relationship.

Median_Household_Income: -0.69023196 - A one-unit increase in the scaled median household income is associated with a decrease of approximately 0.6902 units in the predicted low-income percentage. This suggests a strong negative relationship

Comparison to the Model Before Feature Selection:

The forward selection process successfully reduced the number of features from 13 to 5 while only slightly increasing the MSE and maintaining a high R-squared. This suggests that these 5 selected features are the most important predictors of "Low_Income_Pct" in your dataset.

The magnitudes and signs of the coefficients for the common features ("High_School_Pct", "Median_Home_Value", "Median_Household_Income") are generally consistent in direction (positive or negative relationship) with the coefficients observed in the model with all 13 features. However, the exact magnitudes have changed because the model is now fitted with a different set of predictors. The influence of each variable is considered in the context of the other variables included in the model.

Notice that "Median_Household_Income" still has the largest absolute coefficient magnitude (-0.6902), indicating it remains the strongest predictor of low-income percentage among the selected features.

# Backwards regression

In [ ]:
# backward regression on same data
lreg = LinearRegression()
sfs_backward = sfs(lreg, k_features=5, forward=False, verbose=2, scoring='neg_mean_squared_error')

sfs_backward = sfs_backward.fit(X_train_scaled, y_train)


[2025-04-11 02:33:44] Features: 12/5 -- score: -0.0019387614097626494
[2025-04-11 02:33:44] Features: 11/5 -- score: -0.0019344866658466686
[2025-04-11 02:33:45] Features: 10/5 -- score: -0.0019339150740515544
[2025-04-11 02:33:45] Features: 9/5 -- score: -0.0019345723904779697
[2025-04-11 02:33:45] Features: 8/5 -- score: -0.0019391083281540132
[2025-04-11 02:33:45] Features: 7/5 -- score: -0.0019529469477426987
[2025-04-11 02:33:45] Features: 6/5 -- score: -0.001973116093883215
[2025-04-11 02:33:45] Features: 5/5 -- score: -0.0020516267322730065

In [ ]:
feat_names_back = list(sfs_backward.k_feature_names_)
print(feat_names_back)

['1', '3', '4', '7', '8']


In [ ]:
#map these to feature names
# print features from k_feature_names_
for i in feat_names_back:
    print(features[int(i)])

High_School_Pct
Hispanic_Pct
White_Non_Hispanic_Pct
Median_Home_Value
Median_Household_Income


In [ ]:
# Create a dataframe with only the selected features
selected_columns = [features[int(i)] for i in feat_names_back]
df_selected = df[selected_columns]

# Split the data into train and test sets
X_train, X_test,\
    y_train, y_test = train_test_split(
        df_selected, y,
        test_size=0.2,
        random_state=42)

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Fit a linear regression model using the selected features
reg = LinearRegression()
reg.fit(X_train, y_train)

# Make predictions using the test set
y_pred = reg.predict(X_test)

# Evaluate the model performance
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")
print(f"Intercept: {reg.intercept_}")
print(f"Slope: {reg.coef_}")

Mean Squared Error: 0.0015287895405130065
R-squared: 0.8541029500797266
Intercept: 0.841269958804004
Slope: [-0.28399062 -0.09198038 -0.10411682  0.16832357 -0.69023196]


## output

got the same 5 features as in the forward selection, which is surprising

also the output is identical

THis means:
1. Strong Evidence for the Importance of These 5 Features
2. Potential for a Relatively Stable Model

# Stepwise regression

In [ ]:
# stepwise regression on same data
lreg = LinearRegression()
sfs_step = sfs(lreg, k_features=5, forward=True, floating= True, verbose=2, scoring='neg_mean_squared_error', cv= 5)

In [ ]:
sfs_step = sfs_step.fit(X_train, y_train)


[2025-04-11 02:33:46] Features: 1/5 -- score: -0.0031278321815982725
[2025-04-11 02:33:46] Features: 2/5 -- score: -0.002445076602978731
[2025-04-11 02:33:46] Features: 3/5 -- score: -0.0022459271639572584
[2025-04-11 02:33:46] Features: 4/5 -- score: -0.0021461423835735043
[2025-04-11 02:33:46] Features: 5/5 -- score: -0.0020516267322730065

In [ ]:
feat_names_step = list(sfs_step.k_feature_names_)
print(feat_names_step)

['0', '1', '2', '3', '4']


In [ ]:
for i in feat_names_step:
    print(features[int(i)])

Bachelors_Pct
High_School_Pct
Black_Pct
Hispanic_Pct
White_Non_Hispanic_Pct


In [ ]:
# Create a dataframe with only the selected features
selected_columns = [features[int(i)] for i in feat_names_step]
df_selected = df[selected_columns]

# Split the data into train and test sets
X_train, X_test,\
    y_train, y_test = train_test_split(
        df_selected, y,
        test_size=0.2,
        random_state=42)

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Fit a linear regression model using the selected features
reg = LinearRegression()
reg.fit(X_train, y_train)

# Make predictions using the test set
y_pred = reg.predict(X_test)

# Evaluate the model performance
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")
print(f"Intercept: {reg.intercept_}")
print(f"Slope: {reg.coef_}")

Mean Squared Error: 0.004029948639174444
R-squared: 0.6154097066961394
Intercept: 1.0604610199770543
Slope: [-0.25794653 -0.55652936 -0.05022875 -0.25315585 -0.18215876]


## output

Different Feature Set: Stepwise regression selected a different set of 5 features compared to forward and backward selection. It included "Bachelors_Pct" and "Black_Pct" but excluded "Median_Home_Value" and "Median_Household_Income".

Comparison to previous models: This MSE (0.0040) is noticeably higher than the MSE obtained with the 5 features selected by forward and backward selection (around 0.0015) and also higher than the MSE of the model with all 13 original features (around 0.00146). This indicates that this stepwise-selected model has a considerably larger average prediction error on the test data compared to those models.

Comparison to previous models: This R-squared (0.6154) is lower than the R-squared achieved with the 5 features from forward and backward selection (around 0.8541) and also lower than what might be expected from a model with more relevant predictors. This suggests that this stepwise-selected model captures a smaller proportion of the variance in the target variable.

Worse Performance: The performance metrics (higher MSE and lower R-squared) indicate that this model, selected by stepwise regression, is considerably worse at predicting "Low_Income_Pct" on the test set compared to the model derived from the 5 features identified by forward and backward selection.

Bachelors_Pct: -0.25794653 - A one-unit increase in the scaled percentage of adults with a bachelor's degree is associated with a decrease of approximately 0.2579 units in the predicted low-income percentage.

High_School_Pct: -0.55652936 - A one-unit increase in the scaled percentage of adults with a high school diploma is associated with a decrease of approximately 0.5565 units in the predicted low-income percentage. This has the largest negative impact among these features.

Black_Pct: -0.05022875 - A one-unit increase in the scaled percentage of the Black population is associated with a decrease of approximately 0.0502 units in the predicted low-income percentage. This shows a small negative relationship.

Hispanic_Pct: -0.25315585 - A one-unit increase in the scaled percentage of the Hispanic population is associated with a decrease of approximately 0.2532 units in the predicted low-income percentage.

White_Non_Hispanic_Pct: -0.18215876 - A one-unit increase in the scaled percentage of the non-Hispanic White population is associated with a decrease of approximately 0.1822 units in the predicted low-income percentage.




# formal Model Comparison
Feature Selection Outcome:

Forward and backward selection resulted in the exact same set of five features. This suggests a strong consensus between these two methods regarding the most important predictors when aiming for a 5-feature model.
Stepwise selection yielded a different set of five features, including "Bachelors_Pct" and "Black_Pct" while excluding "Median_Home_Value" and "Median_Household_Income".

Performance Metrics:

MSE: The models from forward and backward selection have a significantly lower MSE (approximately 0.0015) compared to the model from stepwise selection (approximately 0.0040). A lower MSE indicates a better fit to the test data with less average squared error in predictions.1 The stepwise model's MSE is considerably higher, indicating poorer predictive accuracy on the test set.

R-squared: Similarly, the models from forward and backward selection exhibit a much higher R-squared (approximately 0.8541) compared to the stepwise selection model (approximately 0.6154). A higher R-squared signifies that a larger proportion of the variance in the "Low_Income_Pct" is explained by the predictors in the model. The stepwise model explains a substantially smaller portion of the variance.


#Model Evaluation:

Based on these results, we can evaluate the three approaches:

Forward and Backward Selection Models:

These two methods converged to the same optimal set of five features, suggesting these are robustly important predictors for a 5-feature model.

The models built with these features demonstrate strong predictive performance on the test set, as indicated by the low MSE and high R-squared. They explain approximately 85.41% of the variance in the low-income percentage with a relatively low average squared prediction error.

Stepwise Selection Model:

The stepwise selection method identified a different set of features.

The performance metrics (higher MSE and lower R-squared) clearly indicate that this model is inferior to the models obtained from forward and backward selection in terms of predictive accuracy on the test data. It has a larger average prediction error and explains a smaller proportion of the variance in the target variable.

# Conclusion of Comparison and Evaluation:

The models derived from forward and backward feature selection are significantly better than the model derived from stepwise feature selection based on the provided test set performance metrics (MSE and R-squared). The consistency in the selected features between the forward and backward methods further strengthens the case for the importance of "High_School_Pct", "Hispanic_Pct", "White_Non_Hispanic_Pct", "Median_Home_Value", and "Median_Household_Income" as key predictors of "Low_Income_Pct" within a 5-feature linear regression model for this dataset.

The stepwise selection, in this instance, seems to have missed some crucial predictors (like "Median_Home_Value" and "Median_Household_Income") and included others that contribute less to the model's predictive power on the unseen data.




# significance of coefficients

Selected Features and Slope (Coefficients for each scaled selected feature):

High_School_Pct: -0.28399062
Hispanic_Pct: -0.09198038
White_Non_Hispanic_Pct: -0.10411682
Median_Home_Value: 0.16832357
Median_Household_Income: -0.69023196

Median_Household_Income (-0.6902): This has the largest absolute coefficient by far. A one-unit increase in the scaled median household income is associated with a substantial decrease of approximately 0.69 units in the predicted low-income percentage. Given the 0 to 1 scale of the target variable, this is a very large effect, suggesting that median household income is a practically significant predictor of low-income percentage.
**This proves that if your family is poor than you are more likely to be poor**

High_School_Pct (-0.2840): This coefficient is also reasonably large in magnitude. A one-unit increase in the scaled percentage of adults with a high school diploma is associated with a decrease of about 0.28 units in the predicted low-income percentage. This suggests a practically significant negative relationship.
**This makes sense because education leads to higher income and opportunities. The reason it is more easily seen with high school than bachelors is because of how much more common a high school degree is**

Median_Home_Value (0.1683): This coefficient indicates a positive relationship. A one-unit increase in the scaled median home value is associated with an increase of about 0.17 units in the predicted low-income percentage. While smaller than the previous two, this effect could still be practically meaningful depending on the range and variability of the scaled median home value. **This makes sense because when house values are high, it makes it harder to buy homes, a sign that there are more people with low income. Think about big cities where prices of lots are so expensive but there are many homeless people**

White_Non_Hispanic_Pct (-0.1041): A one-unit increase in the scaled percentage of the non-Hispanic White population is associated with a decrease of about 0.10 units in the predicted low-income percentage. This suggests a modest but potentially practically relevant negative relationship. **historically this proves true since generally white people are more well of than minorities**

Hispanic_Pct (-0.0920): This has the smallest absolute coefficient among the selected features. A one-unit increase in the scaled percentage of the Hispanic population is associated with a decrease of about 0.09 units in the predicted low-income percentage. The practical significance of this effect might be smaller compared to the others, but it still indicates a consistent negative relationship. **This is an interesting fact that needs to be understood more**

Assuming a 0-1 Scaling (as mentioned in your initial output): If the scaled features roughly range from 0 to 1, then the coefficients directly indicate the change in "Low_Income_Pct" across the entire normalized range of each predictor.

A change of -0.69 in "Low_Income_Pct" due to the entire scaled range of "Median_Household_Income" is highly practically significant.
A change of -0.28 in "Low_Income_Pct" due to the entire scaled range of "High_School_Pct" is also practically significant.
A change of around 0.10 to 0.17 in "Low_Income_Pct" due to the entire scaled range of "Hispanic_Pct", "White_Non_Hispanic_Pct", and "Median_Home_Value" might be considered moderately practically significant, depending on the context and the decisions being made based on the model.